gnosis_vpn-lib/src/core/disconnection_runner.rs

```diff
@@ -0,0 +1,178 @@
+use edgli::hopr_lib::SessionClientConfig;
+use thiserror::Error;
+use tokio::sync::mpsc;
+
+use std::fmt::{self, Display};
+use std::sync::Arc;
+
+use crate::connection::options::Options;
+use crate::core::disconn::Disconn;
+use crate::core::runner::{self, Results};
+use crate::gvpn_client;
+use crate::hopr::types::SessionClientMetadata;
+use crate::hopr::{Hopr, HoprError};
+use crate::{ping, wg_tooling};
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error(transparent)]
+    Hopr(#[from] HoprError),
+    #[error(transparent)]
+    GvpnClient(#[from] gvpn_client::Error),
+    #[error(transparent)]
+    WgTooling(#[from] wg_tooling::Error),
+    #[error(transparent)]
+    Ping(#[from] ping::Error),
+}
+
+pub struct DisconnectionRunner {
+    disconn: Disconn,
+    hopr: Arc<Hopr>,
+    options: Options,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum Evt {
+    DisconnectWg,
+    OpenBridge,
+    UnregisterWg,
+    CloseBridge,
+}
+
+impl DisconnectionRunner {
+    pub fn new(disconn: Disconn, hopr: Arc<Hopr>, options: Options) -> Self {
+        Self { disconn, hopr, options }
+    }
+
+    pub async fn start(&self, results_sender: mpsc::Sender<Results>) {
+        let res = self.run(results_sender.clone()).await;
+        let _ = results_sender
+            .send(Results::DisconnectionResult {
+                wg_public_key: self.disconn.wg_public_key.clone(),
+                res,
+            })
+            .await;
+    }
+
+    async fn run(&self, results_sender: mpsc::Sender<Results>) -> Result<(), Error> {
+        // 0. disconnect wg tunnel if any
+        if let Some(wg) = self.disconn.wg() {
+            let _ = results_sender
+                .send(Results::DisconnectionEvent {
+                    wg_public_key: self.disconn.wg_public_key.clone(),
+                    evt: Evt::DisconnectWg,
+                })
+                .await;
+            let _ = wg
+                .close_session()
+                .await
+                .map_err(|err| tracing::warn!("disconnecting WireGuard failed: {}", err));
+        }
+
+        let _ = results_sender
+            .send(Results::DisconnectionEvent {
+                wg_public_key: self.disconn.wg_public_key.clone(),
+                evt: Evt::OpenBridge,
+            })
+            .await;
+        // 1. open bridge session
+        let bridge_session = open_bridge_session(&self.hopr, &self.disconn, &self.options).await?;
+
+        // 2. unregister wg public key
+        let _ = results_sender
```

Comment:
Do not ignore the results please...

Reply:
That's nice and dandy, but it's always preferable to log such a situation, since it is not a default or expected code path.
```suggestion
        if let Err(error) = results_sender ...
```